tempesta_fw/sock_srv.c

```diff
@@ -170,29 +169,53 @@ tfw_sock_srv_connect_try(TfwSrvConnection *srv_conn)
 		return r;
 	}
 
+	/*
+	 * Set connection destructor such that connection failover can
+	 * take place if the connection attempt fails.
+	 */
+	conn->destructor = (void *)tfw_srv_conn_release;
+
 	return 0;
 }
 
 static inline void
 tfw_sock_srv_connect_try_later(TfwSrvConnection *srv_conn)
 {
 	/*
-	 * Timeout between connect attempts is increased with each
-	 * unsuccessful attempt. Length of the timeout is decided
-	 * with a variant of exponential backoff delay algorithm.
+	 * Timeout between connect attempts is increased with each unsuccessful
+	 * attempt. Length of the timeout is decided with a variant of
+	 * exponential backoff delay algorithm.
 	 *
-	 * It's essential that the connection is restored as fast
-	 * as possible, so the min retry interval is set to 1. That
-	 * yields the first few reconnect attempts in 1, 2, 4, 8, 16
-	 * milliseconds. The use of a small constant retry interval
-	 * for the first few attempts doesn't yield a better result.
+	 * It's essential that the new connection is established and failed
+	 * connection is restored as fast as possible, so the min retry interval
+	 * is set to 1. The next, second, step is good for loopback
+	 * reconnection, e.g. if upstream is configured to reset connection
+	 * periodically. Following steps are almost pure backoff starting from
+	 * 100ms, good RTT for fast 10Gbps link. We do not increase timeout
+	 * after 1 second as it have moderate overhead and still good in
+	 * response time.
 	 */
-	if (srv_conn->timeout < TFW_SOCK_SRV_RETRY_TIMER_MAX) {
-		srv_conn->timeout = min(TFW_SOCK_SRV_RETRY_TIMER_MAX,
-					TFW_SOCK_SRV_RETRY_TIMER_MIN
-					* (1 << srv_conn->attempts));
-		srv_conn->attempts++;
+	static const unsigned long timeouts[] = { 1, 10, 100, 250, 500, 1000 };
+
+	if (srv_conn->attempts < ARRAY_SIZE(timeouts)) {
+		srv_conn->timeout = timeouts[srv_conn->attempts];
+		TFW_DBG_ADDR("Cannot establish connection",
+			     &srv_conn->conn.peer->addr);
+	} else {
+		srv_conn->timeout = timeouts[ARRAY_SIZE(timeouts) - 1];
+		if (srv_conn->attempts == ARRAY_SIZE(timeouts)
+		    || !(srv_conn->attempts % 60))
+		{
+			char addr_str[TFW_ADDR_STR_BUF_SIZE] = { 0 };
+			tfw_addr_fmt_v6(&srv_conn->conn.peer->addr.v6.sin6_addr,
```

Comment:
I didn't get, why unconditionally format string as ipv6 address? If we connect to server by IPv4, string will be filled with garbage. Better to fill `addr_str` with `tfw_addr_fmt_v6` or `tfw_addr_fmt_v4` chosen by value of `addr.family`